File: src/backend/SyllabusManager.Data/Models/Subjects/Subject.cs
using SyllabusManager.Data.Enums.FieldOfStudies;
using SyllabusManager.Data.Enums.Subjects;
using SyllabusManager.Data.Models.FieldOfStudies;
using SyllabusManager.Data.Models.ManyToMany;
using SyllabusManager.Data.Models.User;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;

namespace SyllabusManager.Data.Models.Subjects
{
    public class Subject : DocumentInAcademicYearBase
    {
        [Required]
        public string Code { get; set; }
        [Required]
        public string NamePl { get; set; }
        public string NameEng { get; set; } = "";
        public ModuleType ModuleType { get; set; }
        public KindOfSubject KindOfSubject { get; set; }
        public MainLanguage Language { get; set; }
        public TypeOfSubject TypeOfSubject { get; set; }
        public SyllabusManagerUser Supervisor { get; set; }
        public List<Literature> Literature { get; set; } = new List<Literature>();
        public List<Lesson> Lessons { get; set; } = new List<Lesson>();
        public List<LearningOutcomeEvaluation> LearningOutcomeEvaluations { get; set; } = new List<LearningOutcomeEvaluation>();
        public List<CardEntries> CardEntries { get; set; } = new List<CardEntries>();
        public List<SubjectTeacher> SubjectsTeachers { get; set; } = new List<SubjectTeacher>();
        [NotMapped]
        public List<SyllabusManagerUser> Teachers { get; set; } = new List<SyllabusManagerUser>();
        public Specialization Specialization { get; set; }
        [NotMapped]
        public bool IsAdmin { get; set; }
        [NotMapped]
        public bool IsSupervisor { get; set; }
        [NotMapped]
        public bool IsTeacher { get; set; }
    }
}

File: src/frontend/src/app/core/models/subject/subject.ts
import { MainLanguage } from "../../enums/field-of-study/main-language.enum";
import { KindOfSubject } from "../../enums/subject/kind-of-subject.enum";
import { ModuleType } from "../../enums/subject/module-type.enum";
import { TypeOfSubject } from "../../enums/subject/type-of-subject.enum";
import { DocumentInAcademicYear } from "../document-in-academic-year";
import { User } from "../user/user";
import { CardEntries } from "./card-entries";
import { LearningOutcomeEvaluation } from "./learning-outcome-evaluation";
import { Lesson } from "./lesson";
import { Literature } from "./literature";

export class Subject extends DocumentInAcademicYear {
    code: string = '';
    namePl: string = '';
    nameEng?: string;
    moduleType: ModuleType = ModuleType.FieldOfStudy;
    kindOfSubject: KindOfSubject = KindOfSubject.InformationTechnology;
    language: MainLanguage = MainLanguage.Polish;
    typeOfSubject: TypeOfSubject = TypeOfSubject.Obligatory;
    discriminator?: string;
    supervisor?: User;
    literature: Literature[] = [];
    lessons: Lesson[] = [];
    learningOutcomeEvaluations: LearningOutcomeEvaluation[] = [];
    cardEntries: CardEntries[] = [];
    teachers: User[] = [];
}
